File: frontend/src/lexer/lexemes.rs
use std::slice::IterMut;

use super::Token;
use super::{Span, WithSpan};

pub type Lexeme = WithSpan<Token>;
pub type LexemesState = usize;

#[derive(Debug)]
pub struct Lexemes {
    lexemes: Vec<Lexeme>,
    offset: usize,
    eof_span: Span,
}

impl<'a> IntoIterator for &'a Lexemes {
    type Item = &'a Lexeme;
    type IntoIter = std::slice::Iter<'a, Lexeme>;

    fn into_iter(self) -> Self::IntoIter {
        self.lexemes.as_slice().into_iter()
    }
}

impl<'a> IntoIterator for &'a mut Lexemes {
    type Item = &'a mut Lexeme;
    type IntoIter = IterMut<'a, Lexeme>;

    fn into_iter(self) -> Self::IntoIter {
        self.lexemes.as_mut_slice().into_iter()
    }
}

impl IntoIterator for Lexemes {
    type Item = Lexeme;
    type IntoIter = std::vec::IntoIter<Lexeme>;

    fn into_iter(self) -> Self::IntoIter {
        self.lexemes.into_iter()
    }
}

impl Lexemes {
    pub fn new(lexemes: Vec<Lexeme>, eof_span: Span) -> Self {
        Self {
            lexemes,
            offset: 0,
            eof_span,
        }
    }

    pub fn is_eof(&self) -> bool {
        self.offset >= self.lexemes.len()
    }

    pub fn peek(&self) -> Lexeme {
        self.peek_nth(0)
    }

    fn eof(&self) -> Lexeme {
        WithSpan::new(Token::EOF, self.eof_span)
    }

    pub fn peek_nth(&self, n: usize) -> Lexeme {
        self.lexemes
            .get(self.offset + n)
            .cloned()
            .unwrap_or(self.eof())
    }

    pub fn peek_n<const N: usize>(&self) -> [Lexeme; N] {
        core::array::from_fn(|i| self.peek_nth(i))
    }

    pub fn next(&mut self) -> Lexeme {
        let res = self.peek();
        self.skip_n(1);
        res
    }

    pub fn skip_n(&mut self, n: usize) {
        self.offset = (self.offset + n).clamp(0, self.lexemes.len());
    }

    pub fn get_state(&self) -> LexemesState {
        self.offset
    }

    pub fn set_state(&mut self, state: LexemesState) {
        self.offset = state;
    }
}

File: src/main/scala/frontend/lexer/Lexemes.scala
package nop.frontend.lexer

import scala.collection.mutable.ArrayBuffer

type Lexeme = WithSpan[Token];

class Lexemes(lexemes: ArrayBuffer[Lexeme], eofSpan: Span):
  private var offset: Int = 0
  private def eof: Lexeme = WithSpan(Token.EOF, eofSpan)

  type State = Int

  /// Check if end of input is reached.
  def isEof: Boolean = offset >= lexemes.length

  /// Peek next lexeme without extracting it. If there's no lexeme, return EOF lexeme.
  def peek: Lexeme = peekNth(0)

  /// Peek `n`th lexeme. If there's not enough lexemes, return EOF lexeme.
  def peekNth(n: Int): Lexeme = lexemes.lift(offset + n).getOrElse(eof)

  /// Peek `n` next lexemes. Any extra lexemes are replaced with EOF lexeme.
  def peekN(n: Int): Seq[Lexeme] = for i <- 0 to n yield peekNth(i)

  /// Get next lexeme. If there's not enough lexemes, return EOF lexeme.
  def next: Lexeme =
    val res = peek
    skipN(1)
    res

  /// Skip `n` next lexemes.
  def skipN(n: Int) =
    offset = (offset + n).min(lexemes.length)

  /// Get lexemes state to save. Used for backtracking.
  def getState: State = offset

  /// Set given lexemes state.
  def setState(state: State) =
    offset = state
